script: Eliminate code duplication in the task queue

mrobinson opened the pull request:
Instead of creating a type for each `TaskSource` variety have each `TaskSource`
hold the same kind of sender (this was inconsistent before, but each
sender was effectively the same trait object), a pipeline, and a
`TaskSourceName`. This elminates the need to reimplement the same
queuing code for every task source.

In addition, have workers hold their own `TaskManager`. This allows just
exposing the manager on the `GlobalScope`. Currently the `TaskCanceller`
is different, but this will also be eliminated in a followup change.

This is a the first step toward having a shared set of `Sender`s on
`GlobalScope`.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes do not require tests because they do not change behavior.

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson committed:
script: Eliminate code duplication in the task queue

Instead of creating a type for each `TaskSource` variety have each `TaskSource`
hold the same kind of sender (this was inconsistent before, but each
sender was effectively the same trait object), a pipeline, and a
`TaskSourceName`. This elminates the need to reimplement the same
queuing code for every task source.

In addition, have workers hold their own `TaskManager`. This allows just
exposing the manager on the `GlobalScope`. Currently the `TaskCanceller`
is different, but this will also be eliminated in a followup change.

This is a the first step toward having a shared set of `Sender`s on
`GlobalScope`.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

mukilan commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -104,7 +103,7 @@ impl ScriptSource for ScriptOrigin {
     final_url: ServoUrl,
     url: ServoUrl,
-    task_source: DOMManipulationTaskSource,
+    task_source: NewTaskSource,
```
Should this be just `TaskSource`?

mrobinson replied:
Thanks for catching this! Yeah, this should be `TaskSource`. This is one reason it isn't great to have commented out code.

mukilan commented on components/script/dom/websocket.rs:
```diff
@@ -168,19 +166,12 @@ impl WebSocket {
+                .global()
+                .task_manager()
+                .websocket_task_source()
+                .queue_unconditionally(BufferedAmountTask { address });
```
Doesn't this change here to use `queue_unconditionally` instead of manually sending the `CommonScriptMsg` mean the Task's `ScriptThreadEventCategory` is now `NetworkEvent` instead of `WebSocketEvent` ? Is that fine? Based on a quick check, now the `WebSocketEvent` is not used anywhere else (except in type functions)

mrobinson replied:
Oh, interesting. I think the right thing here is that the WebSocket task source should be classifying things as `ScriptThreadEventCategory::WebSocket`. I've made this change. I imagine that using `NetworkingEvent` was an oversight before.

mukilan commented on components/script/task_manager.rs:
```diff
@@ -9,35 +9,22 @@ use std::sync::Arc;
-    ($self:ident, $task_source:ident, $task_source_type:ident, $task_source_name:ident) => {
-        pub(crate) fn $task_source(&$self) -> $task_source_type {
+    ($self:ident, $task_source:ident) => {
+        pub(crate) fn $task_source(&$self) -> TaskSource {
```
Not a blocker for this PR, just something I noticed  - in a lot of cases, the task source returned by this method is only ever used for calling `.queue_*()` on it immediately. This method both clones the channel and allocates a new Box and based on adding `println` in the clone method, it seems like this is happening quite a lot. 

Is there a plan to eventually use a single instance of task source, or would that introduce issues related to message ordering?

mrobinson replied:
Yeah, this is pretty inefficient. A lot of times the `TaskSource` is moved into handlers which requires a clone anyway, but it would be nice to only clone when necessary. I hope we can make this change once the canceller is part of the `TaskSource` itself...which should make the API a lot more ergonomic in general.

mrobinson commented:
@mukilan Thanks for the review! I think I've addressed your concerns and also I've renamed `TaskSourceName::Websocket` to `TaskSourceName::WebSocket` for consistency with the rest of the codebase.
